building/build-debs/homeworld-admin-tools/src/setup.py

```diff
@@ -122,18 +122,37 @@ def admit_keyserver(ops: Operations, config: configuration.Config) -> None:
         ops.ssh("confirm that @HOST was admitted", node, "test", "-e", KEYCLIENT_DIR + "/granting.pem")
 
 
-def setup_keygateway(ops: Operations, config: configuration.Config) -> None:
+def setup_keygateway(ops: Operations, config: configuration.Config, overwrite_keytab: bool = False) -> None:
     for node in config.nodes:
         if node.kind != "supervisor":
             continue
         # keytab is stored encrypted in the configuration folder
         keytab = os.path.join(configuration.get_project(), "keytab.%s.crypt" % node.hostname)
         decrypted = keycrypt.gpg_decrypt_to_memory(keytab)
-        ops.ssh("confirm no existing keytab on @HOST", node, "test", "!", "-e", KEYTAB_PATH)
-        ops.ssh_upload_bytes("upload keytab for @HOST", node, decrypted, KEYTAB_PATH)
+        def safe_upload_keytab():
+            if not overwrite_keytab:
+                try:
+                    print("checking for existing keytabs...")
+                    subprocess.check_call(["ssh", "root@%s.%s" % (node.hostname, config.external_domain), "test", "!", "-e", KEYTAB_PATH])
+                except subprocess.CalledProcessError as e_test:
+                    if e_test.returncode != 1:
+                        command.fail(e_test)
+                    print("existing keytab found")
+                    existing_keytab = subprocess.check_output(["ssh", "root@%s.%s" % (node.hostname, config.external_domain), "cat", KEYTAB_PATH])
+                    if existing_keytab != decrypted:
+                        command.fail("existing keytab does not match local keytab")
+                    print("existing keytab matches")
+                    return
```

Comment:
You can probably encode this as a single ssh command, rather than needing to round-trip it multiple times and have all of this helper code involved.

Reply:
Good idea, will implement.